### DISCUSSION TITLE
keystone build --frozen does not exit when schemas are mismatched

### DISCUSSION THREAD

**Discussion by Thinkscape** [keystonejs/keystone#8530] [Questions] [answered]
### Repro steps
1. Having a keystone project with a `graphql.schema` in root dir
2. Modify the keystone schema
3. Run `keystone build --frozen` manually, or as part of `postinstall` script

### Expected

 As [per the documentation](https://keystonejs.com/docs/guides/cli#build-flags) the script should exit with code `> 0` as the schemas are mismatched

### Actual

 The script waits for input and asks if I want to update the schema:
```
> keystone build --frozen

Your Prisma schema is not up to date
✔ Replace the Prisma schema? … 
```

### Env
```
node v18.15.0
@keystone-6/core@5.1.0
```

**Comment by borisno2** [comment]
Hi @Thinkscape 
Currently the `--frozen` flag will check if `process.env.SKIP_PROMPTS` is set, which is automatically set in CI, and if set will `throw new ExitError(1)` otherwise with prompt to update. This might need to be clearer in the documentation.
 
Also, open to feedback if there is a good use case for this to be different. But the general Idea is this Exits with an error in CI but prompts when it is able.

**Comment by dcousens** [comment]
@borisno2 are we using `is-ci` for that?

**Comment by borisno2** [comment]
> @borisno2 are we using `is-ci` for that?

No, we are using `export let shouldPrompt = process.stdout.isTTY && !process.env.SKIP_PROMPTS;` in `pacakges/core/src/lib/prompts.ts`
https://github.com/keystonejs/keystone/blob/818b0b61a1367c39fdc4d2b91f69fe8a4a157adc/packages/core/src/lib/prompts.ts#L31

**Comment by Thinkscape** [comment]
Local dev `postinstall` would not have `SKIP_PROMPTS` present - used for the first build of git-ignored tsgql and prisma client.

I think I know where my thinking is flawed - potentially local dev would use `keystone dev` that should handle the initial build. The intermittent state of IDE not seeing the deps is a minor inconvenience 🤷 

That said, this command seems to be a confusing - would it be used in local dev, or not? 
Given `--frozen` is supposed to be preserving the schemas, I don't understand why it wouldn't do what it says on the tin (and docs). 

I'd either:
* remove this promise from docs and `keystone help` and keep the hidden behaviour,
* or remove the implicit behaviour and depend on existence of `--freeze` for `shouldPrompt`

**Reply by borisno2** [reply, marked as answer]
Thanks for the feedback @Thinkscape

The primary idea of the `--frozen` flag is for use in CI/automated build actions, where you might want it to fail if there is a mismatch between your Keystone config and your schemas - in a similar way to `yarn install --frozen`, but in Keystone's case will prompt if `isTTY` in node is true - See [Node Docs](https://nodejs.org/api/tty.html#tty) for more information - (or `SKIP_PROMPTS` is set - although this one is being removed).

This behaviour was changed in `keystone-6/core@5.0.0` where previously the `keystone build` would behave like `keystone build --frozen` does now - ie validate and exit or prompt if node `isTTY`. The prompting allows this command to be used across both CI and local dev without an exit in dev, for example, `keystone postinstall` is an alias for `keystone build --frozen --no-ui`
